lib/archive/unpack_unix.go

```diff
@@ -131,71 +137,104 @@ func InvokeUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, r
 	return nil
 }
 
-// Unpack hooks into a binary present in the appliance vm called unpack in order to execute InvokeUnpack inside of a chroot. This method works identically to InvokeUnpack, except that it will not function in areas where the binary is not present at /bin/unpack
-func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+// OfflineUnpack wraps Unpack for usage in contexts without a childReaper.
+func OfflineUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+
+	var cmd *exec.Cmd
+	var err error
+	if cmd, err = Unpack(op, tarStream, filter, root, ApplianceBinaryPath); err != nil {
+		return err
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Unpack runs the binary compiled in cmd/unpack.go which creates a chroot at `root` and passes `op`, `tarStream`, and `filter` to InvokeUnpack for extraction of the tar on the filesystem. `binPath` should be either ApplianceBinaryPath or ContainerBinaryPath. Unpack returns a `Cmd` to allow use in conjunction with the tether's `LaunchUtility`, so it is necessary to call `cmd.Wait` after `Unpack` exits e.g. OfflineUnpack, if not being used in conjunction with LaunchUtility and the childReaper.
+func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string, binPath string) (*exec.Cmd, error) {
 
 	fi, err := os.Stat(root)
 	if err != nil {
 		// the target unpack path does not exist. We should not get here.
 		op.Errorf("tar unpack target does not exist: %s", root)
-		return err
+		return nil, err
 	}
 
 	if !fi.IsDir() {
 		err := fmt.Errorf("unpack root target is not a directory: %s", root)
 		op.Error(err)
-		return err
+		return nil, err
 	}
 
 	encodedFilter, err := EncodeFilterSpec(op, filter)
 	if err != nil {
 		op.Error(err)
-		return err
+		return nil, err
 	}
 
 	// Prepare to launch the binary, which will create a chroot at root and then invoke InvokeUnpack
 	// #nosec: Subprocess launching with variable. -- neither variable is user input & both are bounded inputs so this is fine
-	cmd := exec.Command("/bin/unpack", op.ID(), root, *encodedFilter)
+	// "/bin/unpack" on appliance
+	// "/.tether/unpack" inside container
+	cmd := exec.Command(binPath, op.ID(), root, *encodedFilter)
 
-	//stdin
 	stdin, err := cmd.StdinPipe()
 
 	if err != nil {
 		op.Error(err)
-		return err
+		return cmd, err
 	}
 
 	if stdin == nil {
 		err = errors.New("stdin was nil")
 		op.Error(err)
-		return err
+		return cmd, err
 	}
-	done := make(chan error)
+
 	go func() {
-		// copy the tarStream to the binary via stdin; the binary will stream it to InvokeUnpack unchanged
-		defer stdin.Close()
+		if tr, ok := tarStream.(*tar.Reader); ok {
```

Comment:
you definitely may want swap if/else end exit earlier, so code gets more readable here.
Something like:
```
if tr, ok := tarStream.(*tar.Reader); !ok {
    if _, err := io.Copy(stdin, tarStream); err != nil {
        op.Errorf("Error copying tarStream: %s", err.Error())
    }
  return
}
all other logic here.

Reply:
Yeah that's a good point. The structure here is a reflection of the history of the function rather than that of sane design decisions. I'll flip these around, as well as break stuff out into its own function as you suggested in another comment